src/ray/object_manager/common.h

```diff
@@ -92,50 +95,47 @@ struct PlasmaObjectHeader {
   // different data/metadata size.
   uint64_t data_size = 0;
   uint64_t metadata_size = 0;
-
   /// Blocks until all readers for the previous write have ReadRelease'd the
-  /// value. Protects against concurrent writers. Caller must pass consecutive
-  /// versions on each new write, starting with write_version=1.
+  /// value. Protects against concurrent writers.
   ///
-  /// \param write_version The new version for write.
   /// \param data_size The new data size of the object.
   /// \param metadata_size The new metadata size of the object.
   /// \param num_readers The number of readers for the object.
-  void WriteAcquire(int64_t write_version,
-                    uint64_t data_size,
-                    uint64_t metadata_size,
-                    int64_t num_readers);
+  /// \return if the acquire was successful.
+  Status WriteAcquire(uint64_t data_size, uint64_t metadata_size, int64_t num_readers);
 
   /// Call after completing a write to signal that readers may read.
   /// num_readers should be set before calling this.
-  ///
-  /// \param write_version The new version for write. This must match the
-  /// version previously passed to WriteAcquire.
-  void WriteRelease(int64_t write_version);
+  Status WriteRelease();
 
   // Blocks until the given version is ready to read. Returns false if the
   // maximum number of readers have already read the requested version.
   //
   // \param[in] read_version The version to read.
   // \param[out] version_read For normal immutable objects, this will be set to
   // 0. Otherwise, the current version.
-  // \return success Whether the correct version was read and there were still
+  // \return Whether the correct version was read and there were still
   // reads remaining.
-  bool ReadAcquire(int64_t version_to_read, int64_t *version_read);
+  Status ReadAcquire(int64_t version_to_read, int64_t *version_read);
 
   // Finishes the read. If all reads are done, signals to the writer. This is
   // not necessary to call for objects that have num_readers=-1.
   ///
   /// \param read_version This must match the version previously passed in
   /// ReadAcquire.
-  void ReadRelease(int64_t read_version);
+  Status ReadRelease(int64_t read_version);
 #endif
 
   /// Setup synchronization primitives.
   void Init();
 
   /// Destroy synchronization primitives.
   void Destroy();
+
+  /// Helper method to acquire the writer mutex while aborting if the
+  /// error bit is set.
+  /// \return if the mutex was acquired successfully.
+  Status TryAcquireWriterMutex();
```

Comment:
I think it'd be good to add an explicit setter method for setting has_error=true, just for readability.